### DISCUSSION TITLE
How to keep directory in sync when files are deleted/renamed?

### DISCUSSION THREAD

**Discussion by kenliu** [twpayne/chezmoi#2692] [Q&A] [answered]
I'm looking for some way to work with the situation where files are deleted from the source. Occasionally I move files around and it would be good if there were some way to clean up when that happens.

For example, I had a file called `Brewfile` in the root of my home directory and then I decided to move it into a subdirectory in chezmoi, but then had to go manually delete the original file everywhere that my dotfiles are synced with chezmoi.

Another use case is that I have a directory of scripts where I  frequently refactor and delete files while working on them, and because chezmoi doesn't clean up that directory before updating, I'm left with outdated scripts that are deleted from the source but are left in the target.

I could work around this by running a script from chezmoi to delete the target directory at the beginning of every run, but it would be nice if there were a way to tell chezmoi to always delete a specific target directory before applying it or otherwise keep it in sync.

**Comment by kenliu** [comment]
update: I found #1446 that says that chezmoi doesn't know about the history of files and suggests to use `.chezmoiignore` to delete files when moved, which makes sense.

I'm updating my question to ask whether there is some built-in way to tell chezmoi to delete an entire directory before applying as a simple way of keeping a directory in sync.

**Comment by twpayne** [comment]
I think the two things you're looking for are:

[`.chezmoiremove`](https://www.chezmoi.io/reference/special-files-and-directories/chezmoiremove/), as [described in the user guide](https://www.chezmoi.io/user-guide/manage-different-types-of-file/#ensure-that-a-target-is-removed).

The [`exact_` directory attribute](https://www.chezmoi.io/reference/source-state-attributes/).

**Comment by kenliu** [comment, marked as answer]
Thanks! Unfortunately `exact_` seems like it doesn't _quite_ meet my needs because it seems to only work for files in the directory with that attribute and doesn't get applied to subfolders. In my case I'd have to apply the `exact_` attribute to every subfolder which will be somewhat cumbersome.

**Reply by twpayne** [reply]
Then you're likely looking for [`chattr +exact --recursive`](https://www.chezmoi.io/reference/commands/chattr/).

**Reply by kenliu** [reply]
hmm...`chattr` doesn't seem to work with the `--recursive` flag, that seems to be an option on the `add` subcommand.

```
chezmoi chattr +exact install --recursive
chezmoi: unknown flag: --recursive
```

`install` is the name of the target directory

**Reply by twpayne** [reply]
Ah, in the meantime you can use `find` and `xargs` then:

```console
$ find install -type d | sort -r | xargs chezmoi chattr +exact
```

The `sort -r` is needed due to a bug in chezmoi 2.29.2 and earlier. Fix for that and to add support for `--recursive` to `chattr` on the way.

**Reply by kenliu** [reply]
Thank you!
